### PR TITLE
Jax all devices

### PR DESCRIPTION
**Pull request by Thenerdstation**
**Context:**
Previously, only the `default.qubit.jax` device supported the JAX interface. This PR adds JAX interface support to all devices.

**Description of the Change:**
The recent addition of `jax.experimental.host_callback` allows us to finally do `jax -> numpy -> numpy -> jax` within a `jax.jit` function! This PR adds the needed scaffolding to do it. 

**Benefits:**
Everyone can start using JAX more and I won't take any more execuses.

**Possible Drawbacks:**
None.

**Related GitHub Issues:**
#943 

**TODOs**

* Gradient support
* Vmap support

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,24 @@
 
 <h3>New features since last release</h3>
 
+<h4>JAX is now supported on all devices</h4>
+
+* JAX can nw be used as the interface to any device
```
Comment (josh146) [inline .github/CHANGELOG.md:7]:
```suggestion
* JAX can now be used as the interface to any device
```

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,24 @@
...
+  dev = qml.device('cirq.simulator', wires=1)
+  @qml.qnode(dev, interface="jax")
+  def circuit(x):
+      qml.RX(x[1], wires=0)
+      qml.Rot(x[0], x[1], x[2], wires=0)
+      return qml.expval(qml.PauliZ(0))
+
+  weights = jnp.array([0.2, 0.5, 0.1])
+  print(circuit(weights)) # DeviceArray(...)
+
```
Comment (josh146) [inline .github/CHANGELOG.md:21]:
```suggestion
  print(circuit(weights)) # DeviceArray(...)
```

File: .github/CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,24 @@
...
+  ```
 <h3>Improvements</h3>
```
Comment (josh146) [inline .github/CHANGELOG.md:23]:
```suggestion
  ```

<h3>Improvements</h3>
```

File: .github/CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,24 @@
...
+  ```python
+  dev = qml.device('cirq.simulator', wires=1)
+  @qml.qnode(dev, interface="jax")
+  def circuit(x):
+      qml.RX(x[1], wires=0)
+      qml.Rot(x[0], x[1], x[2], wires=0)
+      return qml.expval(qml.PauliZ(0))
+
+  weights = jnp.array([0.2, 0.5, 0.1])
+  print(circuit(weights)) # DeviceArray(...)
```
Comment (josh146) [inline .github/CHANGELOG.md:20]: Woah. This is awesome.

File: pennylane/tape/interfaces/jax.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+
+class JAXInterface(AnnotatedQueue):
+    """Mixin class for applying an JAX interface to a :class:`~.JacobianTape`.
+
+    JAX-compatible quantum tape classes can be created via subclassing:
+
+    .. code-block:: python
+
+        class MyAutogradQuantumTape(AutogradInterface, JacobianTape):
```
Comment (josh146) [inline pennylane/tape/interfaces/jax.py:33]:
```suggestion
        class MyJAXQuantumTape(JAXInterface, JacobianTape):
```

Comment (josh146) [reply pennylane/tape/interfaces/jax.py:33]: P.S. Let's move the interfaces to the device ASAP before the interfaces grow even more 😆

File: pennylane/tape/interfaces/jax.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    Alternatively, the jax interface can be dynamically applied to existing
```
Comment (josh146) [inline pennylane/tape/interfaces/jax.py:35]:
```suggestion
    Alternatively, the JAX interface can be dynamically applied to existing
```

File: pennylane/tape/interfaces/jax.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    quantum tapes via the :meth:`~.apply` class method. This modifies the
+    tape **in place**.
+
+    Once created, the autograd interface can be used to perform quantum-classical
+    differentiable programming.
+
+    .. note::
+
+        If using a device that supports native autograd computation and backpropagation, such as
+        :class:`~.DefaultQubitJAX`, the Autograd interface **does not need to be applied**. It
```
Comment (josh146) [inline pennylane/tape/interfaces/jax.py:45]:
```suggestion
        :class:`~.DefaultQubitJAX`, the JAX interface **does not need to be applied**. It
```

File: pennylane/tape/interfaces/jax.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+    # pylint: disable=attribute-defined-outside-init
+    dtype = jnp.float64
+
+    @property
+    def interface(self):  # pylint: disable=missing-function-docstring
+        return "jax"
+
+    def _execute(self, params, device):
+        # TODO(chase): Add support for this.
```
Comment (josh146) [inline pennylane/tape/interfaces/jax.py:83]: Outdated comment?

File: pennylane/tape/interfaces/jax.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        if len(self.observables) != 1:
+            raise ValueError("Only one return type is supported currently")
+        return_type = self.observables[0].return_type
+        if return_type is not Variance and return_type is not Expectation:
+            raise ValueError(
+                f"Only Variance and Expectation returns are support, given {return_type}"
+            )
+        exec_fn = partial(self.execute_device, device=device)
```
Comment (josh146) [inline pennylane/tape/interfaces/jax.py:91]: Could this be simplified using `self.output_dim`?

```python
return host_callback.call(
    exec_fn, params, result_shape=jax.ShapeDtypeStruct(self.output_dim, JAXInterface.dtype)
)
```

However, there would still need to be one validation; a validation to ensure that there is no mixture of `probs` and `expval`/`var`. E.g., anything that would result in a different output dim to what the tape estimates.

File: pennylane/tape/interfaces/jax.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        ...     qml.RX(0.5, wires=0)
+        ...     expval(qml.PauliZ(0))
+        >>> JAXInterface.apply(tape)
+        >>> tape
+        <JAXQuantumTape: wires=<Wires = [0]>, params=1>
+        """
+        tape_class = getattr(tape, "__bare__", tape.__class__)
+        tape.__bare__ = tape_class
+        tape.__class__ = type("JAXQuantumTape", (cls, tape_class), {})
+        return tape
```
Comment (josh146) [inline pennylane/tape/interfaces/jax.py:116]: In retrospect, this is a design decision I do not like. Even ignoring the dynamic class change, this feels more like a function than a class method?

File: tests/devices/test_default_qubit_jax.py [hunk 11]
```diff
@@ -436,25 +436,6 @@ def cost(weights):
         grad = jax.grad(cost)(weights)
         assert grad.shape == weights.shape
 
-    def test_non_backprop_error(self):
-        """Test that an error is raised in tape mode if the diff method is not backprop"""
```
Comment (josh146) [inline tests/devices/test_default_qubit_jax.py:440]: ❤️

File: tests/tape/interfaces/test_tape_jax.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+                qml.RY(a, wires=0)
+                qml.RX(b, wires=0)
+                qml.expval(qml.PauliZ(0))
+            return tape.execute(device)
+
+        dev = qml.device("default.qubit", wires=1)
+        res = cost(a, b, device=dev)
+        assert res.shape == (1,)
+        # Easiest way to test object is a device array instead of np.array
+        assert "DeviceArray" in res.__repr__()
```
Comment (josh146) [inline tests/tape/interfaces/test_tape_jax.py:69]: :+1:

**Review by josh146 — CHANGES_REQUESTED** [review]
This is awesome, I'm excited to have this in!

Most of my suggestions in the PR itself are minor, but there are some other parts of the codebase that should be modified:

* We likely need an integration test for this new functionality. Could you make a copy of `tests/tape/interfaces/test_qnode_autograd.py` as `test_qnode_jax.py`? Probably the entire file could stay as is, with minor modification to account for the autograd -> JAX change.

* Small tweaks to the [JAX page](https://pennylane.readthedocs.io/en/stable/introduction/interfaces/jax.html) in the docs need to be made to account for this change

File: pennylane/tape/interfaces/jax.py (hunk 5 above)
Comment (Thenerdstation) [reply pennylane/tape/interfaces/jax.py:33]: Agreed.

File: pennylane/tape/interfaces/jax.py (hunk 8 above)
Comment (Thenerdstation) [reply pennylane/tape/interfaces/jax.py:83]: Nope! The comment is for adding support for more than 1 measured observable.

File: pennylane/tape/interfaces/jax.py (hunk 10 above)
Comment (Thenerdstation) [reply pennylane/tape/interfaces/jax.py:116]: I'm usually always against inplace transformations, and this one is pretty intense. tbh I don't really understand how this works.

File: pennylane/tape/interfaces/jax.py (hunk 9 above)
Comment (Thenerdstation) [reply pennylane/tape/interfaces/jax.py:91]: Can I leave that to a future PR where I fix the above todo?

**Comment by Thenerdstation** [comment]
> * We likely need an integration test for this new functionality. Could you make a copy of `tests/tape/interfaces/test_qnode_autograd.py` as `test_qnode_jax.py`? Probably the entire file could stay as is, with minor modification to account for the autograd -> JAX change.

I don't have gradient support yet, so I'll only be able to support some of the tests.

File: pennylane/tape/interfaces/jax.py (hunk 8 above)
Comment (josh146) [reply pennylane/tape/interfaces/jax.py:83]: Oh! In that case, recommend making it more detailed in case other devs come across it
```suggestion
        # TODO (chase): Add support for more than 1 measured observable.
```

File: pennylane/tape/interfaces/jax.py (hunk 9 above)
Comment (josh146) [reply pennylane/tape/interfaces/jax.py:91]: :+1:

File: pennylane/tape/interfaces/jax.py (hunk 10 above)
Comment (josh146) [reply pennylane/tape/interfaces/jax.py:116]: iirc the in-place transform was solely to placate Autograd. Everything else worked at the time without it being in-place (tf, torch), but autograd would give a really bizarre and long traceback that vanished when it was made in-place instead.

Comment (josh146) [reply pennylane/tape/interfaces/jax.py:116]: (it could have been torch that complained rather than autograd, I can't remember fully).

**Comment by Thenerdstation** [comment]
@josh146 Looks like this is still broken despite check on the build passing. Can you tag your github support on this PR?

File: .github/CHANGELOG.md [hunk 13]
```diff
@@ -2,11 +2,23 @@
...
```
Comment (josh146) [inline .github/CHANGELOG.md:20]: I can't suggest it, but this block is missing an end triple-back-quote

Comment (josh146) [reply .github/CHANGELOG.md:20]: https://github.com/PennyLaneAI/pennylane/runs/1855194494#step:4:2022

Comment (Thenerdstation) [reply .github/CHANGELOG.md:20]: Aw thank you. I was wondering why the doc build kept failing here but worked locally.
